[쇼핑 장바구니 1, 2단계] 채넛 미션 제출합니다.

yrsel opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [ ] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?
 
장바구니에서 제거할 때 화면이 업데이트 되는 부분은 완벽하게 구현되지 않아서 참고 부탁드립니다.
구현을 완벽하게 하지 못하다보니 테스트도 작성하지 못했습니다. 추후에 반영하도록 하겠습니다 !
안드로이드 개발에 알고 있었으면 하는 내용 위주로 리뷰 부탁드립니다, 감사합니다 ! 

## 코드 리뷰 커뮤니케이션

안녕하세요, 디노 !
모바일 안드로이드 리뷰이 채넛입니다 
쇼핑 장바구니 미션 잘 부탁드립니다 ! :)
아직 안드로이드를 어떻게 잘 사용해야 하는 지 기준이 잡히지 않아서 부족한 부분 많습니다.
읽기 힘든 코드 보시게 해서 죄송하지만 미션동안 잘 부탁드립니다 !

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

- 실행 화면

https://github.com/user-attachments/assets/3760800b-d51d-4582-a05f-cff25ca613f2

Commit:
docs: README.md 1단계 기능 목록 추가

Commit:
chore: 더미 이미지 리소스 추가

yrsel committed:
build: fragment, databinding 라이브러리 의존성 추가

yrsel committed:
refactor: 상품 목록 아이템 xml 수정

yrsel committed:
feat: Product 상품 도메인 추가

yrsel committed:
build: Glide 라이브러리 의존성 추가

yrsel committed:
chore: 리소스 추가

- colors
- icon

yrsel committed:
build: kapt 의존성 추가

yrsel committed:
feat: BindingAdapter Glide 이미지 확장함수 구현

yrsel committed:
feat: PriceFormatter 금액 단위 표시 변환 구현

yrsel committed:
feat: 더미 데이터 추가

yrsel committed:
feat: ProductRecyclerView 클릭 리스너 구현

yrsel committed:
feat: ProductListAdapter 구현

yrsel committed:
feat: ProductListViewHolder 구현

yrsel committed:
feat: ProductListRecyclerView 아이템 구현

yrsel committed:
feat: ProductListActivity 구현

yrsel committed:
feat: strings.xml 금액 상수화

yrsel committed:
feat: Money 일급 객체 구현

yrsel committed:
chore: Ktlint 적용

Commit:
docs: README.md 업데이트

Commit:
chore: 불필요한 리소스 제거

Commit:
feat: 상품 상세 페이지 xml 구현

Commit:
feat: 상품 상세 페이지 구현

Commit:
feat: Product serialize 직렬화 적용

Commit:
feat: 장바구니 담기 버튼 클릭 리스너 인터페이스 구현

Commit:
refactor: main -> product_list로 네이밍 변경

Commit:
feat: 목록에서 아이템 클릭 시 상세 페이지로 이동

Commit:
chore: ktlint 적용

Commit:
feat: 장바구니 페이지 xml 구현

yrsel committed:
feat: 장바구니 데이터베이스 구현

yrsel committed:
chore: ktlint 적용

Commit:
feat: 장바구니 담기 버튼을 클릭하면 데이터베이스에 저장하는 기능 추가

yrsel committed:
refactor: BindingAdapter로 금액 표시

yrsel committed:
refactor: CartRepository 패키지 변경

- ProductConverter : 상호 변환 확장 함수
- CartRepository 도메인 계층 이동

yrsel committed:
feat: CartActivity 장바구니 페이지 구현

yrsel committed:
refactor: Appbar -> Toolbar 커스텀

yrsel committed:
fix: CartRepository fetch 안되는 현상 수정

- thread 내부에서 콜백 함수 호출

yrsel committed:
feat: ViewModel 생성 팩토리 함수 구현

yrsel committed:
feat: ProductViewModel 뷰모델 생성

Commit:
feat: 상품 상세 페이지 툴바 구현

Commit:
feat: 장바구니 페이지 툴바 구현

Commit:
feat: 삭제 버튼을 누르면 제거하는 기능 추가

Commit:
feat: 상품 목록 페이지에 더보기 버튼 추가

Commit:
refactor: ProductViewModel -> CartViewModel 네이밍 변경

Commit:
chore: 더미 데이터 추가

Commit:
refactor: 패키지 구조 변경

Commit:
feat: ProductRepository 구현

Commit:
feat: ProductListViewModel 구현

Commit:
refactor: 더보기 버튼 아이템으로 분리

Commit:
feat: 상품 목록 더보기 기능 구현

yrsel committed:
chore: 오타 수정

yrsel committed:
refactor: ProductListActivity 더보기 버튼 Span 조절

yrsel committed:
chore: 주석 제거

yrsel committed:
feat: 장바구니 페이지네이션 구현

yrsel committed:
refactor: CartActivity 내부 구조 리팩토링

yrsel committed:
refactor: ProductListActivity 내부 구조 리팩토링

yrsel committed:
chore: ktlint 적용

woowahan-dino commented on app/src/main/java/woowacourse/shopping/data/cart/CartRepositoryImpl.kt:
```diff
@@ -0,0 +1,53 @@
+import woowacourse.shopping.utils.toProductEntity
+import kotlin.concurrent.thread
+
+class CartRepositoryImpl(db: CartDatabase) : CartRepository {
```
실제로 db를 사용하는건 dao뿐이니 CartDatabase대신 CartDao를 생성자에서 인자로 받으면 어떨까요?
이렇게 하면 어떤점이 좋을지도 한번 고민해보세요

yrsel replied:
데이터베이스를 받게된다면 사용할 `DAO` 외에도 데이터베이스에 선언된 모든 `DAO`에 접근할 수 있는 문제가 발생할 것 같습니다.
필요한 `DAO`만 생성자로 주입받는 방식이 `Repository` 마다의 역할을 명확히 구분할 수 있을 것 같습니다 !

관련 커밋
b2f17debe97b9fbd676829469c2c410d33b0a178
129b5ca3e8e99bcf207efd8e6bf56332ef487a4e

woowahan-dino commented on app/src/main/java/woowacourse/shopping/domain/product/ProductRepository.kt:
```diff
@@ -0,0 +1,15 @@
+
+    fun fetchById(id: Long): Product?
+
+    fun canMoreLoad(
```
canMoreLoad보다 canLoadMore가 더 자연스러운 표현이라고 생각합니다

yrsel replied:
페이지네이션에 대해 다시 고민하다가 가져올 데이터보다 1개 더 가져와서 마지막 데이터 인지 비교하는 방식으로 변경해서 해당 메서드는 제거하게 되었습니다.

메서드명 작성할 때 고민했었는데 canLoadMore가 더 자연스러운 표현이었군요, 알려주셔서 감사합니다 디노 ! :)

woowahan-dino commented on app/src/main/java/woowacourse/shopping/domain/product/ProductRepository.kt:
```diff
@@ -0,0 +1,15 @@
+package woowacourse.shopping.domain.product
+
+interface ProductRepository {
+    fun productsByPageNumberAndSize(
```
함수는 동사로 시작하는게 일반적이에요~

yrsel replied:
더미 데이터를 사용하던 방식에서  `product` 테이블을 만들게 되면서 함수명 수정하였습니다.
`fetchXXX`와 같이 동사로 시작하도록 변경하였습니다 !
```kotlin
fun fetchInRange(
    limit: Int,
    offset: Int,
    onResult: (List<Product>) -> Unit,
)
```

woowahan-dino commented on app/src/main/java/woowacourse/shopping/utils/ViewModelFactory.kt:
```diff
@@ -0,0 +1,28 @@
+import woowacourse.shopping.ui.viewmodel.CartViewModel
+import woowacourse.shopping.ui.viewmodel.ProductListViewModel
+
+object ViewModelFactory {
```
이렇게 하는 방식도 좋긴한데 나중에 ViewModel이 엄청 많아지게 되면 관리가 어려울 것 같아요.
각 ViewModel class안에 companion object로 관리하면 어떨까요?

yrsel replied:
한 군데에서 관리하는 것이 유지보수에 이점이 있을 것으로 생각했는데, 디노의 말씀처럼 많아질수록 오히려 관리가 어려울 것 같네요.
좋은 방향 제안해주셔서 감사합니다 ! :) 수정하였습니다 !

관련 커밋 60275fc3411d5b850dde85e092f016060703f6d4

woowahan-dino commented on app/src/main/java/woowacourse/shopping/ui/cart/CartActivity.kt:
```diff
@@ -0,0 +1,114 @@
+            )
+
+        binding.rvCart.adapter = cartAdapter
+        binding.btnPrevious.setOnClickListener { viewModel.moveToPrevious() }
+        binding.btnNext.setOnClickListener { viewModel.moveToNext() }
```
DataBinding가 ViewModel을 받아서 xml에서 처리해보면 어떨까요?

yrsel replied:
xml에서 DataBinding을 활용해서 처리하도록 반영하였습니다 ! :)

관련 커밋 b3fa8ffd247fd03e86fff56bf40a0f2380c52921

woowahan-dino commented on app/src/main/java/woowacourse/shopping/ui/cart/CartAdapter.kt:
```diff
@@ -0,0 +1,41 @@
+
+    @SuppressLint("NotifyDataSetChanged")
+    fun updateItem(it: List<Product>?) {
+        items = it.orEmpty()
```
List자체를 변경하기보다 List안에 있는 요소들만 추가하면 어떨까요?

yrsel replied:
이 부분은 조금 더 알아보고 다음 리뷰에 반영하겠습니다 😭

woowahan-dino commented on app/src/main/java/woowacourse/shopping/ui/productdetail/ProductDetailActivity.kt:
```diff
@@ -0,0 +1,83 @@
+import woowacourse.shopping.ui.productlist.ProductListActivity
+import woowacourse.shopping.utils.intentSerializable
+
+class ProductDetailActivity : AppCompatActivity() {
```
ProductDetail에는 ViewModel이 없네요.
여기에도 ViewModel을 사용해서 구현해볼까요?

yrsel replied:
`ProductDetailActivity` 도 `ViewModel` 사용하여 화면을 보여주도록 수정하였습니다 !

관련 커밋 3b9353ee51ca14acf76c8a6cb44563991bc12dd3

woowahan-dino commented on app/src/main/java/woowacourse/shopping/utils/BindingAdapter.kt:
```diff
@@ -0,0 +1,23 @@
+    @JvmStatic
+    @BindingAdapter("setImageWithGlide")
+    fun ImageView.setImageWithGlide(imageUrl: String) {
+        Glide.with(this.context)
```
with에 Context가 들어가는 것과 View가 들어가는 것이 어떤 차이가 있을까요?

yrsel replied:
다음 리뷰 반영 전 학습 정리 후 남기도록 하겠습니다 🥲

yrsel replied:
`with`에 `View`가 들어가게 된다면 해당 View를 포함하고 있는 Fragment, Activity 생명주기에 연결합니다. View가 아직 attach 되지 않은 상태에서는 작동하지 않으며 가능하면 View 기반이 아닌 Activity, Fragment 기준으로 로드 하는 것을 권장하고 있습니다. 단점으로는 View 계층 구조가 커질수록 비효율적입니다.

`Context`가 들어가게 된다면 애플리케이션 수준의 옵션만 적용되며, 생명주기에 따라 자동으로 시작, 중지되지 않습니다. Context를 넣는 경우는 일반적인 Activity, Fragment 생명주기와 무관하게 리소스를 사용할 경우에 적합합니다.

생명주기에 맞춰 로드하고 관리하고 싶다면 `Context`보다는 해당 생명주기 구성요소를 사용하는 것이 더 적절한 것 같습니다.
두 가지 선택지 중에서 현재 사용 목적에 더 알맞는 것은 생명주기와 연결되는 `View`를 사용하는 것이 더 좋을 것 같습니다.

woowahan-dino commented on app/src/main/java/woowacourse/shopping/utils/BindingAdapter.kt:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @JvmStatic
+    @BindingAdapter("setFormattedPrice")
+    fun TextView.formatted(price: Int) {
+        this.text = this.context.getString(R.string.price, PriceFormatter.format(price))
+    }
```
이정도라면 BindingAdapter를 사용하지 않고 xml에서도 처리가 가능해보이네용
https://stackoverflow.com/questions/38978499/how-do-i-use-databinding-to-combine-a-string-from-resources-with-a-dynamic-varia

yrsel replied:
`PriceFormatter`를 `xml` 마다 주입해줘야다 보니 `@BindingAdapter`로 분리시켰었습니다. 
디노가 말해주신 것처럼 이정도는 `xml`에서 처리하면 좋을 것 같네요!

`strings.xml` 에서 `%,d` 로 천 단위 구분 할 수 있는 사실을 알게되어 `PriceFormatter`클래스를 제거하고 `xml`내부에서 처리하도록 수정하였습니다 ! :)

관련 커밋 195a3dfdc364148ec84942724c355834d74221d5 ( 다른 내용들과 함께 들어가버렸습니다 🥲 )

woowahan-dino reviewed (changes requested):
채넛 안녕하세요.
ui 관련된 기능 구현은 잘 해주셨네요!
추가로 드리고 싶은 코멘트를 남겨볼게요
- Activity이동할 때 이전 액티비티로 돌아가거나, 지금 액티비티를 닫고 다른 액티비티로 가거나 하는 navigate 하는 동작들이 조금 어색한게 있습니다. 앱 사용해보면서 어색하다고 생각하는 부분들을 고쳐보시면 좋을 것 같아요.
- 장바구니 구현이 완성되지 않은 것 같으니 완성하고 다시 리뷰요청 해주세요~

yrsel committed:
refactor: 화면 이동 과정 재정리

yrsel committed:
refactor: ProductDetailActivity 함수 분리 및 순서 정리

- DetailClickListener : SAM(단일 추상 메서드)로 변경

yrsel committed:
refactor: CartActivity 함수 분리 및 순서 정리

- CartClickListener : SAM(단일 추상 메서드) 변경

yrsel committed:
refactor: ProductListActivity 함수 분리 및 순서 정리

- CustomSpanSizeLookup : GridLayout SpanSize 커스텀
- CartClickListener : SAM(단일 추상 메서드) 변경

yrsel committed:
feat: ClothesStoreDatabaseProvider 데이터베이스 제공 객체 생성

yrsel committed:
feat: GlobalApplication 전역적 데이터 초기화 담당 추가

yrsel committed:
refactor: ClothesStoreDatabase 생성

- CartDatabase 제거
- ClothesStoreDatabase 추가

yrsel committed:
refactor: ProductRepository Dao 주입받도록 수정

- 더미데이터 -> 데이터베이스에 더미데이터 추가해서 사용하도록 수정

yrsel committed:
refactor: ProductListViewModel Room DB 활용하는 방식으로 수정

yrsel committed:
feat: CartItemEntity Cart 데이터베이스 관련 객체들 생성

- CartItemEntity : cart 테이블
- CartItemDetail : cart 테이블, product 테이블 조인 객체
- CartProduct : 장바구니 상품 도메인 객체

yrsel committed:
refactor: CartRepository Dao 주입받도록 수정

- CartViewModel : CartProduct 객체의 변경을 감지하도록 수정
- CartDao : Product 데이터베이스 사용 -> product,cart 테이블을 조인하는 방식으로 수정

yrsel committed:
refactor: ViewModelFactory 뷰모델마다 각자 생성해주도록 수정

- ViewModelFactory.kt 제거
- CartViewModel, ProductListViewModel 각자 companion object 로 Factory 변수 제공

yrsel committed:
refactor: CartActivity xml에서 ViewModel 활용해 클릭 리스너 기능 수행

yrsel committed:
refactor: ProductDetailActivity ViewModel 구현

- ProductDetailViewModel : 상품 상세 뷰모델
- DetailClickListener.kt : 제거 -> 뷰모델에서 처리

yrsel committed:
refactor: ProductListAdapter notify 메서드 수정

- notifyDataSetChanged -> notifyItemRangeInserted

yrsel committed:
refactor: CartViewModel 페이지 네이션 적용

yrsel committed:
chore: ktlint 적용

yrsel commented:
> Activity이동할 때 이전 액티비티로 돌아가거나, 지금 액티비티를 닫고 다른 액티비티로 가거나 하는 navigate 하는 동작들이 조금 어색한게 있습니다. 앱 사용해보면서 어색하다고 생각하는 부분들을 고쳐보시면 좋을 것 같아요.

- 화면 이동 로직 수정하여 정상 작동하도록 하였습니다 ! 
  관련 커밋 57d0d88dfc3baf0bc90c732196be2bbc9d28ee73

> 장바구니 구현이 완성되지 않은 것 같으니 완성하고 다시 리뷰요청 해주세요~

- 기능 구현 완료 하였습니다 ! 

리뷰 반영이 많이 늦어져서 죄송합니다 😭

woowahan-dino reviewed (approved):
리뷰 반영 잘 해주셨네요!
3,4 단계 미션을 위해서 1,2 단계 미션은 여기서 마무리하도록 하겠습니다~~
